### PR TITLE
Simplify number parsing

### PR DESCRIPTION
**Pull request by nik9000**
Simplify the code we use to parse numeric values from documents so it is
a bit more readable and easier to share.

### CODE REVIEW

**Comment by nik9000** [comment]
I've opened this as draft so CI can chew on it. The code isn't *literally* the same, but I believe it does all the same things we used to do and I can read it now. My ulterior motive is that I'm building some code that wants to share the value parsing logic and yanking it out like I've done integrates well with my draft. I figured *something* like this change would be useful regardless of how my draft goes.

**Comment by nik9000** [comment]
It looks like this code passes the tests. I'd love some eyes on it. I'm quite open to changing it, but I'd like to get *something* like it in.

File: server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java [hunk 1]
```diff
@@ -1138,42 +1138,39 @@ protected String contentType() {
...
-        }
+    /**
+     * Read the value at the current position of the parser.
+     * @throws InputCoercionException if xcontent couldn't convert the value in the required type, for example, integer overflow
+     * @throws JsonParseException if there was any error parsing the json
+     * @throws IllegalArgumentException if there was an error parsing the value from the json
+     * @throws IOException if there was any other IO error
+     */
+    private static Number value(XContentParser parser, NumberType numberType, Number nullValue, boolean coerce)
+        throws InputCoercionException, JsonParseException, IllegalArgumentException, IOException {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:1165]: Do we ever need to distinguish between the first three exception types here? If not, maybe we should convert them all to a common parsing exception type

**Review by romseygeek — APPROVED** [review]
Makes sense to me. The way we actually handle parsing numbers within XContentParser is a bit weird to me, but that's something to look at separately.

File: server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:1165]: I don't think we need to distinguish, no. I didn't want dig too deep about restructuring them though. I figure just documenting them was a step in the right direction.

**Pull request merged**